File: StriverDP/MinimumPathSum.cpp
/*
https://www.codingninjas.com/codestudio/problems/minimum-path-sum_985349?leftPanelTab=0
*/
#include <bits/stdc++.h> 
int minSumPath(vector<vector<int>> &grid) {
    int n = grid.size();
    int m = grid[0].size();
    vector<int> prev(m,0);
    for(int i = 0 ; i < n ; i++){
        vector<int> curr(m,0);
        for(int j = 0 ;  j < m ; j++){
            if(i==0 && j==0){
                curr[j] = grid[i][j];

            }
            else{
                int up = grid[i][j];
                if(i>0){
                    up += prev[j];
                } 
                else{
                    up += 1e9;
                }
                int left = grid[i][j];
                if(j>0){
                    left += curr[j-1];
                } 
                else{
                    left += 1e9;
                }
                curr[j] = min(up,left);
            }
        }
        prev = curr;
    }
    return prev[m-1];
}

File: StriverDP/minimumPathSum.py
'''
https://www.codingninjas.com/codestudio/problems/minimum-path-sum_985349?leftPanelTab=1
'''

from os import *
from sys import *
from collections import *
from math import *

import sys
sys.setrecursionlimit(10**7)

def minSumPath(grid):
    n = len(grid)
    m = len(grid[0])
    dp = [float('inf') for i in range(m)]
    for i in range(n):
        temp = dp.copy()
        for j in range(m):
            if(i == 0 and j == 0):
                dp[j] = grid[0][0]
            else:
                up = float('inf')
                left = float('inf')
                if(i>0): up = temp[j]
                if(j>0): left = dp[j-1]
                dp[j] = grid[i][j] + min(up,left)
                # print("tp",temp)
                # print("dp",dp)
    return dp[m-1]

# Main.
t = int(input())
while (t > 0):
    l = list(map(int, input().split()))
    n,m = l[0], l[1]
    grid = []
    for i in range(n):
        ll = list(map(int, input().split()))
        grid.append(ll)
    print(minSumPath(grid))
    t -= 1
